test/UnitTestHelpers.hs

```diff
@@ -0,0 +1,360 @@
+  }
+
+testCliStateWithItem :: CliState
+testCliStateWithItem = stubCliState { cliState = testStateWithItem }
+
+-- | Client state mock transformation monad.
+newtype CliMock a = CliMock
+  { runCliMock :: StateT CliState IO a }  -- we build off io so we can compile but we don't want to use it
+  deriving (Monad, Functor, Applicative)
+
+instance MonadStateRead CliMock where
+  {-# INLINE getsState #-}
+  getsState f = CliMock $ gets $ f . cliState
+
+instance MonadStateWrite CliMock where
+  {-# INLINE modifyState #-}
+  modifyState f = CliMock $ state $ \cliS ->
+    let !newCliS = cliS {cliState = f $ cliState cliS}
+    in ((), newCliS)
+  {-# INLINE putState #-}
+  putState newCliState = CliMock $ state $ \cliS ->
+    let !newCliS = cliS {cliState = newCliState}
+    in ((), newCliS)
+
+instance MonadClientRead CliMock where
+  {-# INLINE getsClient #-}
+  getsClient f = CliMock $ gets $ f . cliClient
+  liftIO = CliMock . IO.liftIO
+
+instance MonadClient CliMock where
+  {-# INLINE modifyClient #-}
+  modifyClient f = CliMock $ state $ \cliS ->
+    let !newCliS = cliS {cliClient = f $ cliClient cliS}
+    in ((), newCliS)
+
+-- instance MonadClientSetup CliMock where
+--   saveClient = CliMock $ do
+--     --toSave <- gets cliToSave
+--     cli <- gets cliClient
+--     msess <- gets cliSession
+--     IO.liftIO $ Save.saveToChan toSave (cli, msess)
+
+instance MonadClientUI CliMock where
+  {-# INLINE getsSession #-}
+  getsSession f = CliMock $ gets $ f . fromJust . cliSession
+  {-# INLINE modifySession #-}
+  modifySession f = CliMock $ state $ \cliS ->
+    let !newCliSession = f $ fromJust $ cliSession cliS
+        !newCliS = cliS {cliSession = Just newCliSession}
+    in ((), newCliS)
+  updateClientLeader aid = do
+    s <- getState
+    modifyClient $ updateLeader aid s
+  getCacheBfs = BfsM.getCacheBfs
+  getCachePath = BfsM.getCachePath
+
+-- instance MonadClientReadResponse CliMock where
+--   receiveResponse = CliMock $ do
+--     ChanServer{responseS} <- gets cliDict
+--     IO.liftIO $ takeMVar responseS
+
+-- instance MonadClientWriteRequest CliMock where
+--   sendRequestAI scmd = CliMock $ do
+--     ChanServer{requestAIS} <- gets cliDict
+--     IO.liftIO $ putMVar requestAIS scmd
+--   sendRequestUI scmd = CliMock $ do
+--     ChanServer{requestUIS} <- gets cliDict
+--     IO.liftIO $ putMVar (fromJust requestUIS) scmd
+--   clientHasUI = CliMock $ do
+--     mSession <- gets cliSession
+--     return $! isJust mSession
+
+instance MonadClientAtomic CliMock where
+  {-# INLINE execUpdAtomic #-}
+  execUpdAtomic _ = return ()  -- handleUpdAtomic, until needed, save resources
+    -- Don't catch anything; assume exceptions impossible.
+  {-# INLINE execPutState #-}
+  execPutState = putState
+
+
+-- the compiler is recommending I ditch the testCliState redundancy ... it feels less readable to me that way though ... but maybe that's because I suck at haskell
```

Comment:
I think, the general idea is that, in the long run, less text to read is more readable. If the short text looks strange, learn to read it. You can probably even do `executorCli = runStateT . runCliMock`.

Reply:
If one needs to learn to read it as opposed to just being able to read it by looking at it doesn't that mean it's less readable?
And since less text usually means less information I have trouble believing that either - for example, I would have named 'bla' something like 'makeQuantizedLine' - identifiers are my main tool for documenting programs (a la Michael Feathers, Robert Martin, Steve McConnell)

That said all other things being equal I do like elegant minimal code. In this particular case I can believe that writing idiomatic & elegant Haskell is better.

Reply:
Yes, you are right, point-free notation is less readable to a person that learned to read ordinary programming languages, but didn't learn to read point-free. However, I think it's more readable to a person that learned both. I optimize for myself, which the latter.

I also agree regarding longer names for non-transient identifiers, unconditionally. I'd name it not `bla`, but `BresenhamsLineAlgorithm`, which is the first three words of its haddock. However, I chose to preserve the name from Wikipedia or roguebasin (and probably more generally, from literature and old Fortran code). I think that was a mistake.

Reply:
BTW, if we spent time discussing this, do try `executorCli = runStateT . runCliMock` (in the same single line) and check if that compiles and return in a month and try to read it (though the benefit of succinct code mostly shows when you can move the eye from the code to many pieces important context or of analogous code, and back, many times a second, which also why spurious (nearly-)empty lines hurt).